osquery/core/watcher.cpp

```diff
@@ -306,12 +306,30 @@ void WatcherRunner::watchExtensions() {
   for (const auto& extension : watcher_->extensions()) {
     // Check the extension status, causing a wait.
     int process_status = 0;
-    extension.second->checkStatus(process_status);
+    ProcessState status = extension.second->checkStatus(process_status);
 
-    auto ext_valid = extension.second->isValid();
+    bool ext_valid = (PROCESS_STILL_ALIVE == status);
     auto s = isChildSane(*extension.second);
 
-    if (!ext_valid || (!s.ok() && getUnixTime() >= delayedTime())) {
+    /*
+      Handle 3 cases
+      == case 1:
+      ext_valid == false
+      FLAGS_enable_extensions_watchdog == *
+        - launch new extension
+      == case 2:
+      ext_valid == true
+      FLAGS_enable_extensions_watchdog == false
+        - do nothing
+      == case 3:
+      ext_valid == true
+      FLAGS_enable_extensions_watchdog == true
+        - if unsane then:
+          - kill old extension
+          - launch new extension (edited)
+    */
+    if (!ext_valid || (FLAGS_enable_extensions_watchdog && !s.ok() &&
+                       getUnixTime() >= delayedTime())) {
       if (ext_valid && FLAGS_enable_extensions_watchdog) {
```

Comment:
Maybe this can be simplified and not being nested. If we break out this conditional `ext_valid && !s.ok() && watchdog` then within it set `ext_valid = false`, the next check can be `!ext_valid` and simply create the extension.

We could also exit early if the delay time has not passed.

Reply:
I updated the conditions. Tell me if that's what you meant.
Since I found the code this way easier to understand, I removed the comments above. I am unsure if increased readability isn't a side effect of me looking at this code too long though.